File: sdks/ruby/lib/adriane.rb
# frozen_string_literal: true

require "ffi"

module Adriane
  extend FFI::Library

  ffi_lib ENV.fetch("ADRIANE_C_API_LIB", "adriane_c_api")

  class Result < FFI::Struct
    layout :code, :int,
           :value, :pointer,
           :error, :pointer
  end

  callback :string_callback, [:string, :pointer, :pointer, :pointer], :int
  callback :event_callback, [:string, :pointer], :void

  class Callbacks < FFI::Struct
    layout :user_data, :pointer,
           :on_node, :string_callback,
           :on_condition, :string_callback,
           :on_event, :event_callback
  end

  attach_function :native_engine_version, :adriane_engine_version, [], :pointer
  attach_function :native_validate_graph_json, :adriane_validate_graph_json, [:string], Result.by_value
  attach_function :native_compile_graph_yaml_json, :adriane_compile_graph_yaml_json, [:string], Result.by_value
  attach_function :native_available_providers_json, :adriane_available_providers_json, [], Result.by_value
  attach_function :native_resolve_model_json, :adriane_resolve_model_json, [:string, :pointer, :pointer], Result.by_value
  attach_function :native_list_components_json, :adriane_list_components_json, [], Result.by_value
  attach_function :native_list_prebuilt_json, :adriane_list_prebuilt_json, [], Result.by_value
  attach_function :native_run_component_json, :adriane_run_component_json, [:string, :string, :string], Result.by_value
  attach_function :native_run_prebuilt_json, :adriane_run_prebuilt_json, [:string, :string, :pointer], Result.by_value
  attach_function :native_engine_run_json, :adriane_engine_run_json, [:string, Callbacks.by_value], Result.by_value
  attach_function :native_engine_resume_json, :adriane_engine_resume_json, [:string, Callbacks.by_value], Result.by_value
  attach_function :native_engine_approve_and_resume_json, :adriane_engine_approve_and_resume_json, [:string, Callbacks.by_value], Result.by_value
  attach_function :native_engine_signal_json, :adriane_engine_signal_json, [:string, :string, :string, Callbacks.by_value], Result.by_value
  attach_function :native_engine_replay_json, :adriane_engine_replay_json, [:string, :string, Callbacks.by_value], Result.by_value
  attach_function :native_string_free, :adriane_string_free, [:pointer], :void
  attach_function :native_result_free, :adriane_result_free, [Result.by_value], :void

  module_function

  def engine_version
    ptr = native_engine_version
    raise Error, "adriane_engine_version returned null" if ptr.null?

    ptr.read_string
  ensure
    native_string_free(ptr) if ptr && !ptr.null?
  end

  def validate_graph_json(definition_json)
    unwrap(native_validate_graph_json(definition_json))
  end

  def compile_graph_yaml_json(yaml)
    unwrap(native_compile_graph_yaml_json(yaml))
  end

  def available_providers_json
    unwrap(native_available_providers_json)
  end

  def resolve_model_json(tier, available_json: nil, override_json: nil)
    available = optional_string_pointer(available_json)
    override = optional_string_pointer(override_json)
    unwrap(native_resolve_model_json(tier, available, override))
  end

  def list_components_json
    unwrap(native_list_components_json)
  end

  def list_prebuilt_json
    unwrap(native_list_prebuilt_json)
  end

  def run_component_json(kind, params_json, channels_json)
    unwrap(native_run_component_json(kind, params_json, channels_json))
  end

  def run_prebuilt_json(name, input_json, options_json: nil)
    options = optional_string_pointer(options_json)
    unwrap(native_run_prebuilt_json(name, input_json, options))
  end

  def engine_run_json(spec_json, callbacks)
    unwrap(native_engine_run_json(spec_json, callbacks))
  end

  def engine_resume_json(spec_json, callbacks)
    unwrap(native_engine_resume_json(spec_json, callbacks))
  end

  def engine_approve_and_resume_json(spec_json, callbacks)
    unwrap(native_engine_approve_and_resume_json(spec_json, callbacks))
  end

  def engine_signal_json(spec_json, signal_name, payload_json, callbacks)
    unwrap(native_engine_signal_json(spec_json, signal_name, payload_json, callbacks))
  end

  def engine_replay_json(spec_json, checkpoint_id, callbacks)
    unwrap(native_engine_replay_json(spec_json, checkpoint_id, callbacks))
  end

  def optional_string_pointer(value)
    return FFI::Pointer::NULL if value.nil?

    FFI::MemoryPointer.from_string(value)
  end
  private_class_method :optional_string_pointer

  def unwrap(result)
    if result[:code].zero?
      return result[:value].read_string
    end

    message = result[:error].null? ? "Adriane C API error #{result[:code]}" : result[:error].read_string
    raise Error, message
  ensure
    native_result_free(result)
  end
  private_class_method :unwrap

  class Error < StandardError; end
end

File: sdks/zig/adriane.zig
const std = @import("std");

pub const AdrianeResult = extern struct {
    code: c_int,
    value: ?[*:0]u8,
    err: ?[*:0]u8,
};

pub const AdrianeStringCallback = ?*const fn (
    payload_json: [*:0]const u8,
    user_data: ?*anyopaque,
    value: *?[*:0]const u8,
    err: *?[*:0]const u8,
) callconv(.C) c_int;
pub const AdrianeEventCallback = ?*const fn (payload_json: [*:0]const u8, user_data: ?*anyopaque) callconv(.C) void;

pub const AdrianeCallbacks = extern struct {
    user_data: ?*anyopaque,
    on_node: AdrianeStringCallback,
    on_condition: AdrianeStringCallback,
    on_event: AdrianeEventCallback,
};

extern fn adriane_engine_version() ?[*:0]u8;
extern fn adriane_validate_graph_json(definition_json: [*:0]const u8) AdrianeResult;
extern fn adriane_compile_graph_yaml_json(yaml: [*:0]const u8) AdrianeResult;
extern fn adriane_available_providers_json() AdrianeResult;
extern fn adriane_resolve_model_json(tier: [*:0]const u8, available_json: ?[*:0]const u8, override_json: ?[*:0]const u8) AdrianeResult;
extern fn adriane_list_components_json() AdrianeResult;
extern fn adriane_list_prebuilt_json() AdrianeResult;
extern fn adriane_run_component_json(kind: [*:0]const u8, params_json: [*:0]const u8, channels_json: [*:0]const u8) AdrianeResult;
extern fn adriane_run_prebuilt_json(name: [*:0]const u8, input_json: [*:0]const u8, options_json: ?[*:0]const u8) AdrianeResult;
extern fn adriane_engine_run_json(spec_json: [*:0]const u8, callbacks: AdrianeCallbacks) AdrianeResult;
extern fn adriane_engine_resume_json(spec_json: [*:0]const u8, callbacks: AdrianeCallbacks) AdrianeResult;
extern fn adriane_engine_approve_and_resume_json(spec_json: [*:0]const u8, callbacks: AdrianeCallbacks) AdrianeResult;
extern fn adriane_engine_signal_json(spec_json: [*:0]const u8, signal_name: [*:0]const u8, payload_json: [*:0]const u8, callbacks: AdrianeCallbacks) AdrianeResult;
extern fn adriane_engine_replay_json(spec_json: [*:0]const u8, checkpoint_id: [*:0]const u8, callbacks: AdrianeCallbacks) AdrianeResult;
extern fn adriane_string_free(ptr: ?[*:0]u8) void;
extern fn adriane_result_free(result: AdrianeResult) void;

pub const Error = error{
    NativeError,
    NullVersion,
};

pub fn engineVersion(allocator: std.mem.Allocator) ![]u8 {
    const ptr = adriane_engine_version() orelse return Error.NullVersion;
    defer adriane_string_free(ptr);
    return allocator.dupe(u8, std.mem.span(ptr));
}

pub fn validateGraphJson(allocator: std.mem.Allocator, definition_json: [:0]const u8) ![]u8 {
    return unwrap(allocator, adriane_validate_graph_json(definition_json.ptr));
}

pub fn compileGraphYamlJson(allocator: std.mem.Allocator, yaml: [:0]const u8) ![]u8 {
    return unwrap(allocator, adriane_compile_graph_yaml_json(yaml.ptr));
}

pub fn availableProvidersJson(allocator: std.mem.Allocator) ![]u8 {
    return unwrap(allocator, adriane_available_providers_json());
}

pub fn resolveModelJson(
    allocator: std.mem.Allocator,
    tier: [:0]const u8,
    available_json: ?[:0]const u8,
    override_json: ?[:0]const u8,
) ![]u8 {
    return unwrap(
        allocator,
        adriane_resolve_model_json(
            tier.ptr,
            if (available_json) |value| value.ptr else null,
            if (override_json) |value| value.ptr else null,
        ),
    );
}

pub fn listComponentsJson(allocator: std.mem.Allocator) ![]u8 {
    return unwrap(allocator, adriane_list_components_json());
}

pub fn listPrebuiltJson(allocator: std.mem.Allocator) ![]u8 {
    return unwrap(allocator, adriane_list_prebuilt_json());
}

pub fn runComponentJson(
    allocator: std.mem.Allocator,
    kind: [:0]const u8,
    params_json: [:0]const u8,
    channels_json: [:0]const u8,
) ![]u8 {
    return unwrap(allocator, adriane_run_component_json(kind.ptr, params_json.ptr, channels_json.ptr));
}

pub fn runPrebuiltJson(
    allocator: std.mem.Allocator,
    name: [:0]const u8,
    input_json: [:0]const u8,
    options_json: ?[:0]const u8,
) ![]u8 {
    return unwrap(
        allocator,
        adriane_run_prebuilt_json(name.ptr, input_json.ptr, if (options_json) |value| value.ptr else null),
    );
}

pub fn engineRunJson(allocator: std.mem.Allocator, spec_json: [:0]const u8, callbacks: AdrianeCallbacks) ![]u8 {
    return unwrap(allocator, adriane_engine_run_json(spec_json.ptr, callbacks));
}

pub fn engineResumeJson(allocator: std.mem.Allocator, spec_json: [:0]const u8, callbacks: AdrianeCallbacks) ![]u8 {
    return unwrap(allocator, adriane_engine_resume_json(spec_json.ptr, callbacks));
}

pub fn engineApproveAndResumeJson(allocator: std.mem.Allocator, spec_json: [:0]const u8, callbacks: AdrianeCallbacks) ![]u8 {
    return unwrap(allocator, adriane_engine_approve_and_resume_json(spec_json.ptr, callbacks));
}

pub fn engineSignalJson(
    allocator: std.mem.Allocator,
    spec_json: [:0]const u8,
    signal_name: [:0]const u8,
    payload_json: [:0]const u8,
    callbacks: AdrianeCallbacks,
) ![]u8 {
    return unwrap(allocator, adriane_engine_signal_json(spec_json.ptr, signal_name.ptr, payload_json.ptr, callbacks));
}

pub fn engineReplayJson(
    allocator: std.mem.Allocator,
    spec_json: [:0]const u8,
    checkpoint_id: [:0]const u8,
    callbacks: AdrianeCallbacks,
) ![]u8 {
    return unwrap(allocator, adriane_engine_replay_json(spec_json.ptr, checkpoint_id.ptr, callbacks));
}

fn unwrap(allocator: std.mem.Allocator, result: AdrianeResult) ![]u8 {
    defer adriane_result_free(result);
    if (result.code == 0) {
        const ptr = result.value orelse return allocator.dupe(u8, "");
        return allocator.dupe(u8, std.mem.span(ptr));
    }

    return Error.NativeError;
}
